convert sklearn --> tmva output format included

smoortga opened the pull request.

Commit:
convert sklearn --> tmva output format included

smoortga commented:
@mverzett please check

Commit:
Update sklearn_to_tmva.py

Sorry, forgot to comment out this print statement

mverzett commented on scripts/sklearn_to_tmva.py:
```diff
@@ -0,0 +1,176 @@
+		<Info name="Dir" value="/user_mnt/user/smoortga/CTag/TMVActag_v1/CTaggingCvsB/CMSSW_7_5_1_Retraining"/>
+    	<Info name="Training events" value="1692351"/>
+    	<Info name="TrainingTime" value="7.82964582e+03"/>
+    	<Info name="AnalysisType" value="Classification"/>
```
Do these value matter? If so it would be better to set them dynamically and not hardcoded. What worries me the most are: 
- TMVA/ROOT Release
- Training events

mverzett commented on scripts/sklearn_to_tmva.py:
```diff
@@ -0,0 +1,176 @@
+    	<Option name="IgnoreNegWeightsInTraining" modified="No">False</Option>
+    	<Option name="NTrees" modified="Yes">{ntrees}</Option>
+    	<Option name="MaxDepth" modified="Yes">{maxdepth}</Option>
+    	<Option name="MinNodeSize" modified="No">5%</Option>
```
Actually this may be sometimes wrong in our setting

smoortga replied:
Yes maybe, we should check the definitions of MinNodeSize to our sample split parameter and see if we can somehow save it automatically. The only issue is probably that sklearn uses and absolute number, whereas TMVA uses a percentage

mverzett replied:
Well, we have the total number used for training, so we can always convert. Also modified="No" --> modified="Yes"

mverzett commented on scripts/sklearn_to_tmva.py:
```diff
@@ -0,0 +1,176 @@
+    	<Option name="NTrees" modified="Yes">{ntrees}</Option>
+    	<Option name="MaxDepth" modified="Yes">{maxdepth}</Option>
+    	<Option name="MinNodeSize" modified="No">5%</Option>
+    	<Option name="nCuts" modified="Yes">50</Option>
```
This one too

smoortga replied:
Yes, not sure if there is a parameter for this in sklearn...

mverzett replied:
Does not look like it, so maybe all these parameters do not matter at all but are simply saved by TMVA for bookkeeping

mverzett commented on scripts/sklearn_training_trees.py:
```diff
@@ -213,7 +213,10 @@
+   # *** Sklearn(python)-type training file (.pkl) ***
+   joblib.dump(clf, args.out, compress=True)
+   # *** TMVA-style training file (.xml) ***
+   #convert.gbr_to_tmva(clf,X,TMVAClassification_BDTG.weights.xml,mva_name = "BDTG",coef = 10, var_names = variables)
```
instead of commenting/uncommenting you could do:
1. add a switch that allows saving the weight in xml together with the .pkl (BTW, TMVAClassification_BDTG.weights.xml should go between quotes)
2. Let the system choose, if args.out is "something.pkl" go for pickle, if it is 'something.xml' go for the conversion

smoortga replied:
Ok I agree. I will have a look at it

mverzett commented on scripts/sklearn_training_trees.py:
```diff
@@ -259,6 +262,7 @@
    X_val = rootnp.root2array(extfile.path,'tree',variables,None,0,nfiles_per_sample,args.testEvery,False,'weight')
    X_val = rootnp.rec2array(X_val)
    BDTG =  clf_val.predict_proba(X_val)[:,1]
+   #BDTG = [i*2-1 for i in BDTG] # turn this on when you want to smear the output discriminator over [-1,+1] instead of [0,1]
```
Same here: a switch would be better

Commit:
User input added

--TMVAOut option added to make the interval [-1,1] and Automatically selecting the correct output format from the 'out' extension
